tap_mailchimp/streams.py

```diff
@@ -356,11 +342,41 @@ class ReportEmailActivity(Incremental):
     """Class for 'reports_email_activity' stream"""
     stream_name = 'reports_email_activity'
     extra_fields = ['emails.activity']
-    key_properties = ['campaign_id', 'action', 'email_id', 'timestamp']
+    key_properties = ['_sdc_record_hash']
     streams_to_sync = ['campaigns']
     path = '/reports/{}/email-activity'
     data_key = 'emails'
     replication_keys = ['timestamp']
+    # We must pass a list of fields for which we want data to the Mailchimp API.
+    # As a result, make these fields as automatic, as they are used to generate the '_sdc record hash' Primary Key.
+    extra_automatic_fields = ['campaign_id', 'action', 'email_id', 'timestamp', 'ip']
+
+    def transform_activities(self, records):
+        """Function to move activity at the top-level from the email activity records"""
+
+        for record in records:
+            if 'activity' in record:
+                if '_links' in record:
+                    del record['_links']
+                record_template = dict(record)
+                del record_template['activity']
```

Comment:
What's the intention of these 2 lines?

Reply:
The **reports_email_activity** contains list of activities under the field **activity** along with the fields like campaign_id, email_id at the top level. So, we are looping over the activities and yielding the activities by adding them to the record we already had in the record_template (campaign_id, email_id).